File: backend/src/models/StylePreset.js
const mongoose = require('mongoose');

const StylePresetSchema = new mongoose.Schema({
  id: {
    type: String,
    required: true,
    unique: true,
    index: true
  },
  name: {
    type: String,
    required: true,
    trim: true
  },
  description: {
    type: String,
    required: true,
    trim: true
  },
  category: {
    type: String,
    enum: ['portrait', 'landscape', 'street', 'macro', 'creative', 'low_light'],
    default: 'portrait'
  },
  type: {
    type: String,
    enum: ['built_in', 'user_created', 'community'],
    default: 'built_in'
  },
  visibility: {
    type: String,
    enum: ['public', 'private', 'friends_only'],
    default: 'public'
  },
  isUserGenerated: {
    type: Boolean,
    default: false
  },
  settings: {
    iso: {
      type: mongoose.Schema.Types.Mixed,
      required: true
    },
    aperture: {
      type: String,
      required: true,
      match: /^f\/\d+(\.\d+)?$/
    },
    shutterSpeed: {
      type: String,
      required: true
    },
    whiteBalance: {
      mode: {
        type: String,
        enum: ['auto', 'daylight', 'cloudy', 'tungsten', 'fluorescent', 'flash', 'shade', 'custom'],
        default: 'auto'
      },
      kelvin: {
        type: Number,
        min: 2000,
        max: 10000,
        default: 5500
      },
      shift: {
        magentaGreen: {
          type: Number,
          min: -9,
          max: 9,
          default: 0
        },
        blueAmber: {
          type: Number,
          min: -9,
          max: 9,
          default: 0
        }
      },
      autoWBBias: {
        enabled: {
          type: Boolean,
          default: false
        },
        amber: {
          type: Number,
          min: -3,
          max: 3,
          default: 0
        },
        magenta: {
          type: Number,
          min: -3,
          max: 3,
          default: 0
        }
      },
      priority: {
        type: String,
        enum: ['standard', 'white_priority', 'atmosphere_priority'],
        default: 'standard'
      }
    },
    exposureCompensation: {
      type: String,
      default: '0.0'
    },
    focusMode: {
      type: String,
      enum: ['single', 'continuous', 'manual', 'hyperfocal'],
      default: 'single'
    },
    meteringMode: {
      type: String,
      enum: ['matrix', 'center', 'spot'],
      default: 'matrix'
    },
    colorProfile: {
      type: String,
      default: 'standard'
    }
  },
  thumbnail: {
    type: String,
    default: ''
  },
  tags: [{
    type: String,
    trim: true
  }],
  sceneTypes: [{
    type: String,
    enum: ['portrait', 'landscape', 'macro', 'street', 'architecture', 'nature', 'sports', 'low_light', 'backlight', 'sunset']
  }],
  lightingConditions: [{
    type: String,
    enum: ['bright', 'normal', 'dim', 'very_dark', 'golden_hour', 'blue_hour']
  }],
  compatibility: {
    cameras: [{
      brand: {
        type: String,
        enum: ['canon', 'nikon', 'sony'],
        required: true
      },
      models: [String]
    }]
  },
  metadata: {
    createdBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: function() { return this.isUserGenerated; }
    },
    createdByUsername: {
      type: String,
      required: function() { return this.isUserGenerated && this.visibility === 'public'; }
    },
    originalSettings: {
      type: Object,
      default: null
    },
    shareCode: {
      type: String,
      unique: true,
      sparse: true
    },
    usageCount: {
      type: Number,
      default: 0
    },
    rating: {
      average: {
        type: Number,
        default: 0,
        min: 0,
        max: 5
      },
      count: {
        type: Number,
        default: 0
      }
    },
    featured: {
      type: Boolean,
      default: false
    },
    difficulty: {
      type: String,
      enum: ['beginner', 'intermediate', 'advanced'],
      default: 'beginner'
    }
  },
  isActive: {
    type: Boolean,
    default: true
  }
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

StylePresetSchema.virtual('averageRating').get(function() {
  return this.metadata.rating.average;
});

StylePresetSchema.virtual('totalRatings').get(function() {
  return this.metadata.rating.count;
});

StylePresetSchema.index({ category: 1, type: 1 });
StylePresetSchema.index({ 'metadata.featured': 1, 'metadata.usageCount': -1 });
StylePresetSchema.index({ tags: 1 });
StylePresetSchema.index({ sceneTypes: 1 });

StylePresetSchema.methods.incrementUsage = function() {
  this.metadata.usageCount += 1;
  return this.save();
};

StylePresetSchema.methods.addRating = function(rating) {
  const currentTotal = this.metadata.rating.average * this.metadata.rating.count;
  this.metadata.rating.count += 1;
  this.metadata.rating.average = (currentTotal + rating) / this.metadata.rating.count;
  return this.save();
};

StylePresetSchema.statics.findByScene = function(sceneType) {
  return this.find({
    sceneTypes: sceneType,
    isActive: true
  }).sort({ 'metadata.usageCount': -1 });
};

StylePresetSchema.statics.findByLighting = function(lightingCondition) {
  return this.find({
    lightingConditions: lightingCondition,
    isActive: true
  }).sort({ 'metadata.usageCount': -1 });
};

StylePresetSchema.statics.getFeatured = function() {
  return this.find({
    'metadata.featured': true,
    isActive: true
  }).sort({ 'metadata.usageCount': -1 });
};

StylePresetSchema.statics.getUserPresets = function(userId, includePrivate = false) {
  const query = {
    'metadata.createdBy': userId,
    isActive: true
  };
  
  if (!includePrivate) {
    query.visibility = { $ne: 'private' };
  }
  
  return this.find(query).sort({ createdAt: -1 });
};

StylePresetSchema.statics.getPublicPresets = function() {
  return this.find({
    visibility: 'public',
    isUserGenerated: true,
    isActive: true
  }).sort({ 'metadata.usageCount': -1 });
};

StylePresetSchema.statics.findByShareCode = function(shareCode) {
  return this.findOne({
    'metadata.shareCode': shareCode,
    isActive: true
  });
};

StylePresetSchema.methods.generateShareCode = function() {
  if (!this.metadata.shareCode) {
    this.metadata.shareCode = Math.random().toString(36).substring(2, 12).toUpperCase();
  }
  return this.metadata.shareCode;
};

StylePresetSchema.methods.canUserAccess = function(userId) {
  // Public presets are accessible to everyone
  if (this.visibility === 'public') return true;
  
  // Private presets only accessible to owner
  if (this.visibility === 'private') {
    return this.metadata.createdBy && this.metadata.createdBy.toString() === userId.toString();
  }
  
  // TODO: Implement friends_only logic when friend system is added
  if (this.visibility === 'friends_only') {
    return this.metadata.createdBy && this.metadata.createdBy.toString() === userId.toString();
  }
  
  return false;
};

// White Balance helper methods
StylePresetSchema.methods.getEffectiveWBKelvin = function() {
  const modeKelvinMap = {
    'tungsten': 3200,
    'fluorescent': 4000,
    'daylight': 5500,
    'flash': 5500,
    'cloudy': 6500,
    'shade': 7500
  };

  if (this.settings.whiteBalance.mode === 'custom') {
    return this.settings.whiteBalance.kelvin;
  }
  
  return modeKelvinMap[this.settings.whiteBalance.mode] || 5500;
};

StylePresetSchema.methods.getWBShiftDescription = function() {
  const { magentaGreen, blueAmber } = this.settings.whiteBalance.shift;
  
  if (magentaGreen === 0 && blueAmber === 0) {
    return 'No shift';
  }
  
  const mgDesc = magentaGreen > 0 ? `M${magentaGreen}` : magentaGreen < 0 ? `G${Math.abs(magentaGreen)}` : '';
  const baDesc = blueAmber > 0 ? `A${blueAmber}` : blueAmber < 0 ? `B${Math.abs(blueAmber)}` : '';
  
  return [mgDesc, baDesc].filter(Boolean).join(', ') || 'No shift';
};

StylePresetSchema.methods.isWBCompatibleWith = function(cameraModel) {
  // Different cameras have different WB shift ranges and precision
  const cameraSpecs = {
    'canon': { mgRange: [-9, 9], baRange: [-9, 9], precision: 1 },
    'nikon': { mgRange: [-6, 6], baRange: [-6, 6], precision: 1 },
    'sony': { mgRange: [-9, 9], baRange: [-9, 9], precision: 1 },
    'fujifilm': { mgRange: [-9, 9], baRange: [-9, 9], precision: 1 }
  };
  
  const brand = cameraModel.toLowerCase().includes('canon') ? 'canon' :
               cameraModel.toLowerCase().includes('nikon') ? 'nikon' :
               cameraModel.toLowerCase().includes('sony') ? 'sony' :
               cameraModel.toLowerCase().includes('fuji') ? 'fujifilm' : 'canon';
  
  const spec = cameraSpecs[brand];
  const wb = this.settings.whiteBalance;
  
  return wb.shift.magentaGreen >= spec.mgRange[0] && 
         wb.shift.magentaGreen <= spec.mgRange[1] &&
         wb.shift.blueAmber >= spec.baRange[0] && 
         wb.shift.blueAmber <= spec.baRange[1];
};

module.exports = mongoose.model('StylePreset', StylePresetSchema);

File: mobile/lib/models/style_preset.dart
class StylePreset {
  final String id;
  final String name;
  final String description;
  final String category;
  final String type;
  final CameraSettings settings;
  final String thumbnail;
  final List<String> tags;
  final List<String> sceneTypes;
  final List<String> lightingConditions;
  final PresetMetadata metadata;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;

  StylePreset({
    required this.id,
    required this.name,
    required this.description,
    required this.category,
    required this.type,
    required this.settings,
    this.thumbnail = '',
    this.tags = const [],
    this.sceneTypes = const [],
    this.lightingConditions = const [],
    required this.metadata,
    this.isActive = true,
    required this.createdAt,
    required this.updatedAt,
  });

  factory StylePreset.fromJson(Map<String, dynamic> json) {
    return StylePreset(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      category: json['category'] ?? 'portrait',
      type: json['type'] ?? 'built_in',
      settings: CameraSettings.fromJson(json['settings'] ?? {}),
      thumbnail: json['thumbnail'] ?? '',
      tags: List<String>.from(json['tags'] ?? []),
      sceneTypes: List<String>.from(json['sceneTypes'] ?? []),
      lightingConditions: List<String>.from(json['lightingConditions'] ?? []),
      metadata: PresetMetadata.fromJson(json['metadata'] ?? {}),
      isActive: json['isActive'] ?? true,
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'category': category,
      'type': type,
      'settings': settings.toJson(),
      'thumbnail': thumbnail,
      'tags': tags,
      'sceneTypes': sceneTypes,
      'lightingConditions': lightingConditions,
      'metadata': metadata.toJson(),
      'isActive': isActive,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  StylePreset copyWith({
    String? id,
    String? name,
    String? description,
    String? category,
    String? type,
    CameraSettings? settings,
    String? thumbnail,
    List<String>? tags,
    List<String>? sceneTypes,
    List<String>? lightingConditions,
    PresetMetadata? metadata,
    bool? isActive,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return StylePreset(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      category: category ?? this.category,
      type: type ?? this.type,
      settings: settings ?? this.settings,
      thumbnail: thumbnail ?? this.thumbnail,
      tags: tags ?? this.tags,
      sceneTypes: sceneTypes ?? this.sceneTypes,
      lightingConditions: lightingConditions ?? this.lightingConditions,
      metadata: metadata ?? this.metadata,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  String toString() {
    return 'StylePreset(id: $id, name: $name, category: $category)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is StylePreset && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;
}

class CameraSettings {
  final dynamic iso;
  final String aperture;
  final String shutterSpeed;
  final String whiteBalance;
  final String exposureCompensation;
  final String focusMode;
  final String meteringMode;
  final String colorProfile;

  CameraSettings({
    required this.iso,
    required this.aperture,
    required this.shutterSpeed,
    this.whiteBalance = 'auto',
    this.exposureCompensation = '0.0',
    this.focusMode = 'single',
    this.meteringMode = 'matrix',
    this.colorProfile = 'standard',
  });

  factory CameraSettings.fromJson(Map<String, dynamic> json) {
    return CameraSettings(
      iso: json['iso'] ?? 400,
      aperture: json['aperture'] ?? 'f/4.0',
      shutterSpeed: json['shutterSpeed'] ?? '1/125',
      whiteBalance: json['whiteBalance'] ?? 'auto',
      exposureCompensation: json['exposureCompensation'] ?? '0.0',
      focusMode: json['focusMode'] ?? 'single',
      meteringMode: json['meteringMode'] ?? 'matrix',
      colorProfile: json['colorProfile'] ?? 'standard',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'iso': iso,
      'aperture': aperture,
      'shutterSpeed': shutterSpeed,
      'whiteBalance': whiteBalance,
      'exposureCompensation': exposureCompensation,
      'focusMode': focusMode,
      'meteringMode': meteringMode,
      'colorProfile': colorProfile,
    };
  }

  CameraSettings copyWith({
    dynamic iso,
    String? aperture,
    String? shutterSpeed,
    String? whiteBalance,
    String? exposureCompensation,
    String? focusMode,
    String? meteringMode,
    String? colorProfile,
  }) {
    return CameraSettings(
      iso: iso ?? this.iso,
      aperture: aperture ?? this.aperture,
      shutterSpeed: shutterSpeed ?? this.shutterSpeed,
      whiteBalance: whiteBalance ?? this.whiteBalance,
      exposureCompensation: exposureCompensation ?? this.exposureCompensation,
      focusMode: focusMode ?? this.focusMode,
      meteringMode: meteringMode ?? this.meteringMode,
      colorProfile: colorProfile ?? this.colorProfile,
    );
  }

  String get displayISO {
    return 'ISO $iso';
  }

  String get displayAperture {
    return aperture;
  }

  String get displayShutterSpeed {
    return shutterSpeed;
  }

  @override
  String toString() {
    return 'CameraSettings(iso: $iso, aperture: $aperture, shutterSpeed: $shutterSpeed)';
  }
}

class PresetMetadata {
  final String? createdBy;
  final int usageCount;
  final PresetRating rating;
  final bool featured;
  final String difficulty;

  PresetMetadata({
    this.createdBy,
    this.usageCount = 0,
    required this.rating,
    this.featured = false,
    this.difficulty = 'beginner',
  });

  factory PresetMetadata.fromJson(Map<String, dynamic> json) {
    return PresetMetadata(
      createdBy: json['createdBy'],
      usageCount: json['usageCount'] ?? 0,
      rating: PresetRating.fromJson(json['rating'] ?? {}),
      featured: json['featured'] ?? false,
      difficulty: json['difficulty'] ?? 'beginner',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'createdBy': createdBy,
      'usageCount': usageCount,
      'rating': rating.toJson(),
      'featured': featured,
      'difficulty': difficulty,
    };
  }

  PresetMetadata copyWith({
    String? createdBy,
    int? usageCount,
    PresetRating? rating,
    bool? featured,
    String? difficulty,
  }) {
    return PresetMetadata(
      createdBy: createdBy ?? this.createdBy,
      usageCount: usageCount ?? this.usageCount,
      rating: rating ?? this.rating,
      featured: featured ?? this.featured,
      difficulty: difficulty ?? this.difficulty,
    );
  }
}

class PresetRating {
  final double average;
  final int count;

  PresetRating({
    this.average = 0.0,
    this.count = 0,
  });

  factory PresetRating.fromJson(Map<String, dynamic> json) {
    return PresetRating(
      average: (json['average'] ?? 0.0).toDouble(),
      count: json['count'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'average': average,
      'count': count,
    };
  }

  PresetRating copyWith({
    double? average,
    int? count,
  }) {
    return PresetRating(
      average: average ?? this.average,
      count: count ?? this.count,
    );
  }

  bool get hasRatings => count > 0;

  String get displayRating {
    if (!hasRatings) return 'No ratings';
    return '${average.toStringAsFixed(1)} ($count rating${count == 1 ? '' : 's'})';
  }
}

// Preset categories enum
enum PresetCategory {
  portrait('portrait', 'Portrait'),
  landscape('landscape', 'Landscape'),
  street('street', 'Street'),
  macro('macro', 'Macro'),
  creative('creative', 'Creative'),
  lowLight('low_light', 'Low Light');

  const PresetCategory(this.value, this.displayName);
  final String value;
  final String displayName;

  static PresetCategory fromString(String value) {
    return PresetCategory.values.firstWhere(
      (category) => category.value == value,
      orElse: () => PresetCategory.portrait,
    );
  }
}

// Scene types enum
enum SceneType {
  portrait('portrait', 'Portrait'),
  landscape('landscape', 'Landscape'),
  macro('macro', 'Macro'),
  street('street', 'Street'),
  architecture('architecture', 'Architecture'),
  nature('nature', 'Nature'),
  sports('sports', 'Sports'),
  lowLight('low_light', 'Low Light'),
  backlight('backlight', 'Backlight'),
  sunset('sunset', 'Sunset');

  const SceneType(this.value, this.displayName);
  final String value;
  final String displayName;

  static SceneType fromString(String value) {
    return SceneType.values.firstWhere(
      (type) => type.value == value,
      orElse: () => SceneType.portrait,
    );
  }
}

// Lighting conditions enum
enum LightingCondition {
  bright('bright', 'Bright'),
  normal('normal', 'Normal'),
  dim('dim', 'Dim'),
  veryDark('very_dark', 'Very Dark'),
  goldenHour('golden_hour', 'Golden Hour'),
  blueHour('blue_hour', 'Blue Hour');

  const LightingCondition(this.value, this.displayName);
  final String value;
  final String displayName;

  static LightingCondition fromString(String value) {
    return LightingCondition.values.firstWhere(
      (condition) => condition.value == value,
      orElse: () => LightingCondition.normal,
    );
  }
}